### PR TITLE
Display logmessage if two nodes run on different versions 

### LINKED ISSUE #8075: Log if two corresponding nodes run on different versions of Icinga 2
## Is your feature request related to a problem? Please describe.

During debugging of customer setups we sometimes stumbled upon environments where one of the nodes was merely forgotten when updates were rolled out. It's not very common, to be honest, but it would help if we could get that option off the table without going through a full diagnostics output of all corresponding nodes.

## Describe the solution you'd like

I'd like to see in the Icinga log when the communiction partner in an api connection runs on a different version of Icinga 2. And I'd like to see this message repeated once in a while because sometimes you can't get debug logs and sometimes customers just send a short timeframe of logs. Especially when it comes to huge setups.

## Describe alternatives you've considered

* Asking for versions. If a node is forgotten, it's forgotten when you ask explicitely, too
* Asking for Icinga Diagnostics outputs. It helps but it's a lot of data to look at

### PR DESCRIPTION
**Pull request by Yonas-net**
resolves #8075

### CODE REVIEW

File: lib/remote/apilistener.cpp [hunk 1]
```diff
@@ -129,6 +131,7 @@ void ApiListener::OnConfigLoaded()
 {
 	if (m_Instance)
 		BOOST_THROW_EXCEPTION(ScriptError("Only one ApiListener object is allowed.", GetDebugInfo()));
+#include <lib/base/logger.hpp>
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:134]: Srsly?

File: lib/remote/apilistener.cpp [hunk 2]
```diff
@@ -1643,6 +1644,11 @@ Value ApiListener::HelloAPIHandler(const MessageOrigin::Ptr& origin, const Dicti
 
 			if (endpoint) {
 				endpoint->SetIcingaVersion((double)params->Get("version"));
+				if ((double)params->Get("version") < (double)l_AppVersionInt - 100) {
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1647]: Make sure `(double)params->Get("version")` is called only once.

File: lib/remote/apilistener.cpp [hunk 3]
```diff
@@ -1642,10 +1642,11 @@ Value ApiListener::HelloAPIHandler(const MessageOrigin::Ptr& origin, const Dicti
 			auto endpoint (client->GetEndpoint());
 
 			if (endpoint) {
-				endpoint->SetIcingaVersion((double)params->Get("version"));
-				if ((double)params->Get("version") < (double)l_AppVersionInt - 100) {
+				double appVersion = params->Get("version");
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1645]: Make it of the same datatype as `l_AppVersionInt`.

File: lib/remote/apilistener.cpp [hunk 4]
```diff
@@ -1607,6 +1635,24 @@ std::set<HttpServerConnection::Ptr> ApiListener::GetHttpClients() const
...
 {
+	if (origin) {
+		auto client (origin->FromClient);
+
+		if (client) {
+			auto endpoint (client->GetEndpoint());
+
+			if (endpoint) {
+				unsigned long appVersion = params->Get("version");
+				endpoint->SetIcingaVersion(appVersion);
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1646]: Add a newline after this one.

File: lib/remote/apilistener.cpp [hunk 5]
```diff
@@ -1607,6 +1635,24 @@ std::set<HttpServerConnection::Ptr> ApiListener::GetHttpClients() const
...
+				if (appVersion < (double)l_AppVersionInt - 100) {
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1647]: What's the cast for?

File: lib/remote/apilistener.cpp [hunk 6]
```diff
@@ -1644,14 +1644,18 @@ Value ApiListener::HelloAPIHandler(const MessageOrigin::Ptr& origin, const Dicti
...
-						<< ". Expected one of: " << l_AppVersionInt << ", " << l_AppVersionInt - 100;
+				
+				if (endpoint->GetZone()->GetParent() == Zone::GetLocalZone()) {
+					switch (l_AppVersionInt / 100 - nodeVersion / 100) {
+						case 0:
+						case 1:
+							break;
+						default:
+							Log(LogWarning, "ApiListener")
+								<< "Unexpected Icinga version of endpoint '" << endpoint->GetName() << "': " << nodeVersion
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1655]: This version has three digits...

File: lib/remote/apilistener.cpp [hunk 7]
```diff
@@ -1644,14 +1644,18 @@ Value ApiListener::HelloAPIHandler(const MessageOrigin::Ptr& origin, const Dicti
...
+								<< ". Expected one of: " << l_AppVersionInt << ", " << l_AppVersionInt - 100;
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1656]: ... but these have five.

File: lib/remote/apilistener.cpp [hunk 8]
```diff
@@ -1633,6 +1633,14 @@ std::set<HttpServerConnection::Ptr> ApiListener::GetHttpClients() const
 	return m_HttpClients;
 }
 
+static void PrintAppVersion(unsigned long nodeVersion, Log& log)
+{
+    log << nodeVersion / 100u / 100u << "." << nodeVersion / 100u % 100u << ".x"
+        << " . Expected one of: " << l_AppVersionInt / 100u / 100u << "."
+        << l_AppVersionInt / 100u % 100u << ".x" << ", " << (l_AppVersionInt / 100u - 1u) / 100u
+        << "." << (l_AppVersionInt / 100u - 1u) % 100u << ".x";
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1641]: Print only `nodeVersion`...

File: lib/remote/apilistener.cpp [hunk 9]
```diff
@@ -1651,12 +1659,8 @@ Value ApiListener::HelloAPIHandler(const MessageOrigin::Ptr& origin, const Dicti
...
 							break;
 						default:
-							Log(LogWarning, "ApiListener")
-								<< "Unexpected Icinga version of endpoint '" << endpoint->GetName() << "': "
-								<< nodeVersion / 100u / 100u << "." << nodeVersion / 100u % 100u << ".x"
-								<< " . Expected one of: " << l_AppVersionInt / 100u / 100u << "."
-								<< l_AppVersionInt / 100u % 100u << ".x" << ", " << (l_AppVersionInt / 100u - 1u) / 100u
-								<< "." << (l_AppVersionInt / 100u - 1u) % 100u << ".x";
+							PrintAppVersion(nodeVersion,Log(LogWarning, "ApiListener")
+								<< "Unexpected Icinga version of endpoint '" << endpoint->GetName() << "': ");
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1663]: ... and call it here as often as you need it.

File: lib/remote/apilistener.cpp [hunk 10]
```diff
@@ -1651,12 +1659,8 @@ Value ApiListener::HelloAPIHandler(const MessageOrigin::Ptr& origin, const Dicti
 						case 1:
 							break;
 						default:
-							Log(LogWarning, "ApiListener")
-								<< "Unexpected Icinga version of endpoint '" << endpoint->GetName() << "': "
-								<< nodeVersion / 100u / 100u << "." << nodeVersion / 100u % 100u << ".x"
-								<< " . Expected one of: " << l_AppVersionInt / 100u / 100u << "."
-								<< l_AppVersionInt / 100u % 100u << ".x" << ", " << (l_AppVersionInt / 100u - 1u) / 100u
-								<< "." << (l_AppVersionInt / 100u - 1u) % 100u << ".x";
+							PrintAppVersion(nodeVersion,Log(LogWarning, "ApiListener")
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1662]: Make `Log(LogWarning, "ApiListener")` a variable `Log log (LogWarning, "ApiListener");` and continue with `log << ...`.

File: lib/remote/apilistener.cpp [hunk 11]
```diff
@@ -1659,8 +1656,15 @@ Value ApiListener::HelloAPIHandler(const MessageOrigin::Ptr& origin, const Dicti
...
-								<< "Unexpected Icinga version of endpoint '" << endpoint->GetName() << "': ");
+							Log log (LogWarning, "ApiListener");
+							log << "Unexpected Icinga version of endpoint '" << endpoint->GetName() << "': ";
+
+							LogAppVersion(nodeVersion / 100u, log);
+							log << " . Expected one of: ";
+
+							LogAppVersion(l_AppVersionInt / 100u, log);
+							log << ", " << (l_AppVersionInt / 100u - 1u) / 100u
+								<< "." << (l_AppVersionInt / 100u - 1u) % 100u << ".x";
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1667]: Why don't you use the function also here?

File: lib/remote/apilistener.cpp [hunk 12]
```diff
@@ -1605,8 +1633,45 @@ std::set<HttpServerConnection::Ptr> ApiListener::GetHttpClients() const
...
+					switch (l_AppVersionInt / 100 - nodeVersion / 100) {
+						case 0:
+						case 1:
+							break;
+						default:
+							Log log (LogWarning, "ApiListener");
+							log << "Unexpected Icinga version of endpoint '" << endpoint->GetName() << "': ";
+
+							LogAppVersion(nodeVersion / 100u, log);
+							log << " . Expected one of: ";
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1663]: You don't need a . here.

**Comment by Yonas-net** [comment]
Testprotocol:
So far it was the problem that if a node runs on 2 versions behind or 1 version before its parent, you could not see any log messages that could indicate this. But this bug has now been fixed and I will illustrate it to you as follows.

Let us assume that the node is version 2.10.x and the master 2.12.x should then see a log message, because it is not allowed to use node that is 2 versions behind master.

Master Version ``v2.12.0`` and node Version ``v2.10.0``.
Output:
```
warning/ApiListener: Unexpected Icinga version of endpoint 'sat': 2.10.x  Expected one of: 2.12.x, 2.11.x
```
However, if the node has the same version as the master or is 1 version behind the master, we do not need a log message because it is allowed.

The master can be 1 version before the node, but the node cannot be 1 version before the master.
```
[2020-07-07 16:40:39 +0200] warning/ApiListener: Unexpected Icinga version of endpoint 'sat': 2.13.x . Expected one of: 2.12.x, 2.11.x
```

**Comment by Al2Klimov** [comment]
> lto1: internal compiler error: in prune_unused_types_prune, at dwarf2out.c:22275

**Comment by Al2Klimov** [comment]
If you ask me, that error is pretty silly.

@lazyfrosch Any chance they fix it for CentOS 7?

**Comment by lazyfrosch** [comment]
Uhm sorry not my expertise

**Comment by lazyfrosch** [comment]
GCC bugs can occur, especially with older maintained versions. We should find a workaround to avoid the problem.

**Comment by Al2Klimov** [comment]
This is likely #7149.

**Comment by Al2Klimov** [comment]
## Blocked by

* [ ] #7905

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Rebase with -i. Keep only your commit.

File: lib/remote/apilistener.cpp [hunk 13]
```diff
@@ -1652,8 +1657,31 @@ Value ApiListener::HelloAPIHandler(const MessageOrigin::Ptr& origin, const Dicti
...
+
+							LogAppVersion(nodeVersion / 100u, log);
+							log << " Expected one of: ";
+
+							LogAppVersion(l_AppVersionInt / 100u, log);
+							log << ", ";
+
+							LogAppVersion((l_AppVersionInt / 100u - 1u), log);
+					}
+				}
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:1684]: Why did you remove SetCapabilities()?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: lib/remote/apilistener.cpp (hunk 13 above)
Comment (Yonas-net) [reply lib/remote/apilistener.cpp:1684]: It was not part of my commit.

Comment (Al2Klimov) [reply lib/remote/apilistener.cpp:1684]: Come on. Only your commits are here.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Re-test https://github.com/Icinga/icinga2/pull/8088#issuecomment-654911523 (just to be sure).

**Comment by Yonas-net** [comment]
Tested it and it works fine!

```powershell
[2021-01-11 08:51:25 +0100] information/ApiListener: Finished sending config file updates for endpoint 'satellite' in zone 'satellite'.
[2021-01-11 08:51:25 +0100] warning/ApiListener: Unexpected Icinga version of endpoint 'satellite': 2.10.x Expected one of: 2.12.x, 2.11.x
```

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**